File: examples/c/gmconfig_c/gmconfig.c
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/** 
 * @file gmconfig.c
 *
 * A C example demonstration of the GMSEC publisher.  This example loads 
 * connection and message information from a standard config file.
 *  
 */

#include "../example.h"


typedef struct
{
  GMSEC_Status     status;
  GMSEC_Config     config;
  GMSEC_ConfigFile config_file;
  GMSEC_Connection connection;
  GMSEC_Message    message;
  GMSEC_SubscriptionInfo info;
} gmconfig_t;


int gmconfig_Initialize(gmconfig_t* this, const char* cfgFilename)
{
	this->status      = statusCreate();
	this->config_file = configFileCreate();
	this->config      = NULL;
	this->connection  = NULL;
	this->message     = NULL;

	/* Load the configuration file and parse it */
	configFileLoad(this->config_file, cfgFilename, this->status);
	if (!example_check("configFileLoad", this->status))
	{
		fprintf(stderr, "%s\n", statusGet(this->status));
		return GMSEC_FALSE;
	}

	/* Get the configuration from the file */
	this->config = configFileLookupConfig(this->config_file, "config1", this->status);
	if (!example_check("configFileLookupConfig", this->status))
	{
		fprintf(stderr, "%s\n", statusGet(this->status));
		return GMSEC_FALSE;
	}

	/* Initilaize the configuration */
	example_initialize(this->config);

	return GMSEC_TRUE;
}


int gmconfig_Run(gmconfig_t *this)
{
	int k = 0;
	const char* pattern = NULL;

	/* Output the GMSEC API version */
	GMSEC_INFO("%s", connectionGetAPIVersion());

	/* Dump the configuration */
	GMSEC_INFO("Using configuration:\n%s", configToXML(this->config, this->status));

	/* Establish the Connection */
	this->connection = example_openConnection(this->config, this->status);
	if (this->connection == NULL) return GMSEC_FALSE;

	/* Output the middleware version */
	GMSEC_INFO("Middleware version: %s", connectionGetLibraryVersion(this->connection, this->status));

	/* Lookup subject pattern by name */
	pattern = configFileLookupSubscription(this->config_file, "events", this->status);
	if (!example_check("configFileLookupSubscription", this->status)) 
	{
		return GMSEC_FALSE;
	}

	GMSEC_INFO("Subscribing to: %s", pattern);

	/* Subscribe on the connection */
	this->info = connectionSubscribe(this->connection, pattern, this->status);
	if (!example_check("connectionSubscribe", this->status)) return GMSEC_FALSE;

	/* Get Message from the configuration file */
	this->message = configFileLookupMessage(this->config_file, "msg1", this->status);
	if (!example_check("configFileLookupMessage", this->status)) return GMSEC_FALSE;

	for (k=0; k < 10; ++k)
	{
		/* Display XML representation of the message */
		GMSEC_INFO("Prepared:\n%s", messageToXML(this->message, this->status));

		/* Publish Message */
		connectionPublish(this->connection, this->message, this->status);
		if (!example_check("Publish(connection)", this->status)) return GMSEC_FALSE;

		GMSEC_INFO("Message published!");
	}

	return GMSEC_TRUE;
}


void gmconfig_Cleanup(gmconfig_t *this)
{
	/* Destroy the Config */
	if (this->config != NULL)
	{
		configDestroy(&this->config);
	}

	/* Destroy the ConfigFile */
	if (this->config_file != NULL)
	{
		configFileDestroy(&this->config_file);
	}

	/* Destroy the Message */
	if (this->message != NULL)
	{
		messageDestroy(&(this->message));
	}

	GMSEC_INFO("Unsubscribing from %s", subscriptionInfoGetSubject(this->info));
	connectionUnsubscribe(this->connection, &(this->info), this->status);
	if (!example_check("Unsubscribing...", this->status))
	{
		GMSEC_ERROR("Problem with Unsubscribing...");
	}

	/* Destroy the connection */
	if (this->connection != NULL)
	{
		example_closeConnection(this->connection, this->status);
	}

	/* Destroy the Status */
	statusDestroy(&(this->status));
}


int main(int argc, char* argv[])
{
	gmconfig_t gmconfig;

	if (argc <= 1)
	{
		printf("usage: gmconfig_c <filename>\n");
		return -1;
	}

	if (!gmconfig_Initialize(&gmconfig, argv[1]))
	{
		fprintf(stderr, "\nFailure of initializing the example");
		return -1;
	}

	gmconfig_Run(&gmconfig);
	gmconfig_Cleanup(&gmconfig);

	return 0;
}

File: examples/perl/gmconfig.pl
#
# Copyright 2007-2016 United States Government as represented by the
# Administrator of The National Aeronautics and Space Administration.
# No copyright is claimed in the United States under Title 17, U.S. Code.
# All Rights Reserved.
#

# 
# gmconfig.pl
# 
# A perl example demonstration of loading a configuration from a standard
# xml file.  The configuration file used for the example code is "gmconfig.xml".
# 


use strict;
use Util;

use libgmsec_perl;

*isa = \&UNIVERSAL::isa;

{
	my $self = bless({});

	# If no arguments were provided, print usage and exit with an erroneous code value
	if (scalar(@ARGV) < 1)
	{
		$self->printUsage();
		exit -1;
	}

	# Initialize logging in the GMSEC API in order to display output
	$self->init();
	
	eval
	{
		$self->run($ARGV[0]);
	};
	if ($@)
	{
		print "caught: $@";
	}
}


sub printUsage
{
	print "\nusage: perl gmconfig.pl <filename>\n";
}


sub init
{
	my $config = libgmsec_perl::Config->new();
	$config->addValue("LOGLEVEL", "INFO");
}


sub run
{
	my ($self, $cfgFilename) = @_;

	# Attempt to load configuration file
	libgmsec_perl::LogInfo("Attempting to load config: " . $cfgFilename);
	my $cfgFile = libgmsec_perl::ConfigFile->new();
	$cfgFile->load($cfgFilename);

	# Attempt to retrieve configurations from the configuration file
	my $config1 = $cfgFile->lookupConfig("config1");
	my $config2 = $cfgFile->lookupConfig("config2");

	# Display log of XML representation of configuration objects
	libgmsec_perl::LogInfo("Config 1:\n" . $config1->toXML());
	libgmsec_perl::LogInfo("Config 2:\n" . $config2->toXML());

	# Attempt to create a Connection object using a configuration object
	my $connection = libgmsec_perl::Connection::create($config2);

	# Destroy the connection object
	libgmsec_perl::Connection::destroy($connection);

	# Attempt to lookup subscription topic from configuration file
	my $pattern = $cfgFile->lookupSubscription("events");

	if (!$pattern)
	{
		die libgmsec_perl::Exception->new($libgmsec_perl::CONFIGFILE_ERROR,
						$libgmsec_perl::OTHER_ERROR_CODE,
						"Failed to acquire 'events' subscription topic");
	}

	libgmsec_perl::LogInfo("Subscription topic: " . $pattern);

	# Attempt to lookup message from configuration file
	my $message = $cfgFile->lookupMessage("msg1");

	# Display XML representation of the message
	libgmsec_perl::LogInfo("Message:\n" . $message->toXML());

	# Obtain Configuration File Iterator
	my $iter = $cfgFile->getIterator();

	# Acquire and display all configuration entries
	while ($iter->hasNextConfig())
	{
		my $configEntry = $iter->nextConfig();

		libgmsec_perl::LogInfo("\nConfig Name: " . $configEntry->getName()
			   . "\nConfig:\n" . $configEntry->getConfig()->toXML());
	}

	# Acquire and display all message entries
	while ($iter->hasNextMessage())
	{
		my $messageEntry = $iter->nextMessage();

		libgmsec_perl::LogInfo("\nMessage Name: " . $messageEntry->getName()
			   . "\nMessage:\n" . $messageEntry->getMessage()->toXML());
	}

	# Acquire and display all subscription entries
	while ($iter->hasNextSubscription())
	{
		my $subscriptionEntry = $iter->nextSubscription();

		libgmsec_perl::LogInfo("\nSubscription Name : " . $subscriptionEntry->getName()
			   . "\nSubscription Topic: " . $subscriptionEntry->getSubject());
	}

	# Acquire and display all custom XML entries
	while ($iter->hasNextCustomElement())
	{
		my $customElement = $iter->nextCustomElement();

		libgmsec_perl::LogInfo("\nCustom XML Element:\n" . $customElement);
	}
}


1;

__END__
